### PR TITLE
chore: Add new, optional top level SDK data system configuration

### PR DESCRIPTION
**Pull request by keelerm84**

### CODE REVIEW

File: servicedef/sdk_config.go [hunk 1]
```diff
@@ -104,3 +105,47 @@ type SDKConfigWrapper struct {
 	Name    string `json:"name"`
 	Version string `json:"version"`
 }
+
+type SDKConfigDataSystem struct {
+	Synchronizers []SDKConfigDataSystemDataSource `json:"synchronizers,omitempty"`
```
Comment (keelerm84) [inline servicedef/sdk_config.go:110]: @cwaldren-ld I know we talked about only having a primary and secondary. It seems reasonable to me to model it as an array for future expansion, but enforce that order by convention. It also allows opting in to secondary support nice as it doesn't really change the shape. You just process the next item on the list in order.

File: servicedef/sdk_config.go [hunk 2]
```diff
@@ -104,3 +105,47 @@ type SDKConfigWrapper struct {
...
+
+type DataSystemDataSourceConfig interface{}
+
+type DataSystemStreamingDataSourceConfig struct {
+	SDKConfigStreamingParams
+}
+
+type DataSystemPollingDataSourceConfig struct {
+	SDKConfigPollingParams
+}
```
Comment (keelerm84) [inline servicedef/sdk_config.go]: I wasn't sure the best way to do this in this language, so figured I should just ask you @cwaldren-ld.

Basically we want to have the 2 separate types of configs (streaming vs polling), but I also want a key telling me which type it is.

How would you structure this?

File: servicedef/sdk_config.go [hunk 3]
```diff
@@ -104,3 +105,47 @@ type SDKConfigWrapper struct {
...
+
+type SDKConfigDataSystemDataSource struct {
+	Type string `json:"type"`
+	DataSystemDataSourceConfig
+}
+
+type DataSystemDataSourceConfig interface{}
+
+type DataSystemStreamingDataSourceConfig struct {
+	SDKConfigStreamingParams
```
Comment (keelerm84) [inline servicedef/sdk_config.go]: Embedding the existing polling and streaming configs we are defining and using above.

File: servicedef/sdk_config.go (hunk 1 above)
Comment (cwaldren-ld) [reply servicedef/sdk_config.go:110]: Let's go with this and see if the "shape" works fine for SDks that implement the test.

File: servicedef/sdk_config.go [hunk 4]
```diff
@@ -104,3 +105,55 @@ type SDKConfigWrapper struct {
...
+	Initializers  []SDKConfigDataSystemDataSource `json:"initializers,omitempty"`
+	Persistence   SDKConfigDataSystemPersistence  `json:"persistence,omitempty"`
+}
+
+type SDKConfigDataSystemDataSource struct {
+	Type string `json:"type"`
+	DataSystemDataSourceConfig
+}
+
+type DataSystemDataSourceConfig interface{}
```
Comment (cwaldren-ld) [inline servicedef/sdk_config.go:120]: What is this line doing?

File: servicedef/sdk_config.go [hunk 5]
```diff
@@ -104,3 +105,55 @@ type SDKConfigWrapper struct {
...
+
+type SDKConfigDataSystemPersistence struct {
+	Store SDKConfigDataSystemPersistenceStore `json:"store"`
+	Cache SDKConfigDataSystemPersistenceCache `json:"cache"`
+}
+
+type SDKConfigDataSystemPersistenceType string
+
+const (
+	Redis    SDKConfigDataSystemPersistenceType = "redis"
```
Comment (cwaldren-ld) [inline servicedef/sdk_config.go]: Another style of doing this is `Redis = SDKConfigDataSystemPersistenceType("redis")`

I think they are equivalent.

File: servicedef/sdk_config.go [hunk 6]
```diff
@@ -104,3 +105,55 @@ type SDKConfigWrapper struct {
...
+
+const (
+	Off      SDKConfigDataSystemPersistenceMode = "off"
+	TTL      SDKConfigDataSystemPersistenceMode = "ttl"
+	Infinite SDKConfigDataSystemPersistenceMode = "infinite"
+)
+
+type SDKConfigDataSystemPersistenceCache struct {
+	Mode string `json:"mode"`
+	TTL  int    `json:"ttl"` // This value is only valid when the Mode is set to TTL. It must be a positive integer.
```
Comment (cwaldren-ld) [inline servicedef/sdk_config.go]: I think this could be `o.Maybe[int]`, with tag `json:"ttl,omitempty`

File: servicedef/sdk_config.go (hunk 2 above)
Comment (cwaldren-ld) [reply servicedef/sdk_config.go]: I think the way we've handled this before is two different fields.

```golang
type Thing struct {
       Type string `json:"type"`
       OptionA o.Maybe[OptionA] `json:"optionA,omitempty"`
       OptionB o.Maybe[OptionB] `json:"optionB,omitempty"`
}
```

For example, our Command config (just using C++ for convenience on my part)
```
struct CommandParams {
    Command command; <-- enum
    std::optional<EvaluateFlagParams> evaluate;
    std::optional<EvaluateAllFlagParams> evaluateAll;
    std::optional<CustomEventParams> customEvent;
    std::optional<IdentifyEventParams> identifyEvent;
    std::optional<ContextBuildParams> contextBuild;
    std::optional<ContextConvertParams> contextConvert;
    CommandParams();
};
```

Comment (keelerm84) [reply servicedef/sdk_config.go]: Updated to match your suggested format.

File: servicedef/sdk_config.go (hunk 5 above)
Comment (keelerm84) [reply servicedef/sdk_config.go]: Updated.

File: servicedef/sdk_config.go (hunk 6 above)
Comment (keelerm84) [reply servicedef/sdk_config.go]: Done.

**Review by cwaldren-ld — APPROVED** [review]

**Pull request merged**